clash-prelude/src/Clash/Sized/Vector/Par.hs

```diff
@@ -0,0 +1,85 @@
+{-|
+Copyright  :  2022     , QBayLogic B.V.
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+-}
+
+{-# LANGUAGE GADTs #-}
+
+{-# LANGUAGE Trustworthy #-}
+
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise       #-}
+
+{-# OPTIONS_GHC -fno-warn-incomplete-patterns #-}
+
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+module Clash.Sized.Vector.Par ( scanlPar, scanrPar ) where
+
+import Prelude hiding ((!!), (++), length, map)
+
+import GHC.TypeLits (KnownNat, type (^), sameNat)
+
+import Clash.Promoted.Nat (SNat (..), SNatLE (..), compareSNat)
+import Clash.Sized.Vector
+import Clash.Sized.RTree (RTree (..), v2t)
+import Data.Type.Equality ((:~:) (..))
+import Data.Proxy (Proxy (..))
+
+{- $setup
+>>> :m -Prelude
+>>> import Clash.Prelude
+-}
+
+-- | Low-depth (left) scan
+--
+-- >>> scanlPar (+) (1 :> 2 :> 3 :> 4 :> Nil)
+-- 1 :> 3 :> 6 :> 10 :> Nil
+scanlPar :: KnownNat n
+         => (a -> a -> a) -- ^ Must be commutative
+         -> Vec (2^n) a -> Vec (2^n) a
+scanlPar op v = scanlInductiveRTree op (v2t v)
+{-# INLINE scanlPar #-}
+
+-- | Low-depth (right) scan
+--
+-- >>> scanrPar (+) (1 :> 2 :> 3 :> 4 :> Nil)
+-- 10 :> 9 :> 7 :> 4 :> Nil
+scanrPar :: KnownNat n
+         => (a -> a -> a) -- ^ Must be commutative
+         -> Vec (2^n) a -> Vec (2^n) a
+scanrPar op v = scanrInductiveRTree op (v2t v)
+{-# INLINE scanrPar #-}
+
+scanlInductiveRTree
+  :: forall a n. KnownNat n
+  => (a -> a -> a)
+  -> RTree n a
+  -> Vec (2^n) a
+scanlInductiveRTree op tr =
+  -- I have to use sameNat and compareSNat both; otherwise it doesn't compile
```

Comment:
This comment is true but doesn't really add why it doesn't compile. Just write something like: "sameNat and compareSNat are both required because type checker doesn't know `n <= 0 <=> n ~ 0`"